### PR TITLE
Add drugZ as a module

### PR DESCRIPTION
**Pull request by LaurenceKuhl**
## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/crisprseq/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, also make a PR on the nf-core/crisprseq _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Comment by LaurenceKuhl** [comment]
@nf-core-bot fix linting

**Comment by LaurenceKuhl** [comment]
I'm not sure why pre-commit is leading to an error :/

File: docs/usage/screening.md [hunk 1]
```diff
@@ -103,16 +103,29 @@ If there are several designs to be run, you can input a folder containing all th
...
 BAGEL2 uses the same contrasts from `--contrasts`.
 
-### MAGECKFlute
+### Running drugZ
 
-The downstream analysis involves distinguishing essential, non-essential, and target-associated genes. Additionally, it encompasses conducting biological functional category analysis and pathway enrichment analysis for these genes. Furthermore, it provides visualization of genes within pathways, enhancing user exploration of screening data. MAGECKFlute is run automatically after MAGeCK MLE and for each MLE design matrice. If you have used the `--day0_label`, MAGeCKFlute will be ran on all the other conditions. Please note that the DepMap data is used for these plots.
+DrugZ detects synergistic and suppressor drug-gene interactions in CRISPR screens. DrugZ is an open-source Python software for the analysis of genome-scale drug modifier screens. The software accurately identifies genetic perturbations that enhance or suppress drug activity. To run drugZ, you can specify `--drugz` which resembles the previous contrast file :
```
Comment (mashehu) [inline docs/usage/screening.md:119]:
```suggestion
DrugZ detects synergistic and suppressor drug-gene interactions in CRISPR screens. DrugZ is an open-source Python software for the analysis of genome-scale drug modifier screens. The software accurately identifies genetic perturbations that enhance or suppress drug activity. To run drugZ, you can specify `--drugz` which resembles the previous contrast file:
```
🇫🇷🙅

File: modules/local/drugz.nf [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args       = task.ext.args ?: ''
+    def prefix     = task.ext.prefix ?: "${meta.id}"
+    def output     = "${meta.treatment}_vs_${meta.reference}_drugz_output.txt".replaceAll(',','_')
+    def foldchange = "${meta.treatment}_vs_${meta.reference}.foldchange".replaceAll(',','_')
+
+    """
+    drugz.py -i $count_table -o $output -f $foldchange -c $meta.reference -x $meta.treatment $args
```
Comment (mashehu) [inline modules/local/drugz.nf:30]: Any way to make this a conda recipe?

File: nextflow_schema.json [hunk 3]
```diff
@@ -230,6 +235,14 @@
                     "type": "string",
                     "description": "Non essential gene set  for BAGEL2",
                     "default": "https://raw.githubusercontent.com/hart-lab/bagel/master/NEGv1.txt"
+                },
+                "drugz": {
+                    "type": "string",
```
Comment (mashehu) [inline nextflow_schema.json:240]:
```suggestion
                    "type": "string",
                    "format":"file-path",
```

File: nextflow_schema.json [hunk 4]
```diff
@@ -230,6 +235,14 @@
...
+                    "description": "Specifies drugz to be run and your contrast file on which comparisons should be done"
+                },
+                "drugz_remove_genes": {
+                    "type": "string",
+                    "description": "Essential genes to remove from the drugZ modules"
```
Comment (mashehu) [inline nextflow_schema.json:245]: Maybe add a pattern to catch spaces etc.

File: workflows/crisprseq_screening.nf [hunk 5]
```diff
@@ -36,6 +36,7 @@ include { paramsSummaryMultiqc                         } from '../subworkflows/n
 include { softwareVersionsToYAML                       } from '../subworkflows/nf-core/utils_nfcore_pipeline'
 include { methodsDescriptionText                       } from '../subworkflows/local/utils_nfcore_crisprseq_pipeline'
 include { validateParametersScreening                  } from '../subworkflows/local/utils_nfcore_crisprseq_pipeline'
+include { DRUGZ                       } from '../modules/local/drugz'
```
Comment (mashehu) [inline workflows/crisprseq_screening.nf:39]: Not harshil formatted

File: workflows/crisprseq_screening.nf [hunk 6]
```diff
@@ -294,6 +296,19 @@ workflow CRISPRSEQ_SCREENING {
         }
     }
 
+    // Launch module drugZ
+    if(params.drugz) {
+        Channel.fromPath(params.drugz)
+                .splitCsv(header:true, sep:';' )
```
Comment (mashehu) [inline workflows/crisprseq_screening.nf:302]: Wait, you said comma separated, now you look for a semi-colon?

**Comment by mashehu** [comment]
@nf-core-bot fix linting

File: workflows/crisprseq_screening.nf (hunk 6 above)
Comment (LaurenceKuhl) [reply workflows/crisprseq_screening.nf:302]: I updated the documentation to make it more clear

File: modules/local/drugz.nf (hunk 2 above)
Comment (LaurenceKuhl) [reply modules/local/drugz.nf:30]: https://github.com/nf-core/crisprseq/issues/189 made it into an issue, might try this during the hackathons

File: bin/BAGEL.py [hunk 7]
```diff
@@ -667,18 +666,18 @@ def calculate_bayes_factors(
...
-    if no_resampling == True:
+    if no_resampling:
         print("# Caution: Resampling is disabled")
         LOOPCOUNT = 1
 
     print("Iter TrainEss TrainNon TestSet")
     sys.stdout.flush()
     for loop in range(LOOPCOUNT):
-        currentbf = {}
+        # currentbf = {}
```
Comment (mashehu) [inline bin/BAGEL.py:680]:
```suggestion
```

File: bin/BAGEL.py [hunk 8]
```diff
@@ -987,7 +984,7 @@ def calculate_bayes_factors(
         #
 
         for loop in range(LOOPCOUNT):
-            currentnbf = {}
+            # currentnbf = {}
```
Comment (mashehu) [inline bin/BAGEL.py:987]:
```suggestion
```

File: bin/drugz.py [hunk 9]
```diff
@@ -123,7 +123,7 @@ def calculate_fold_change(
     fc_replicate_id = "fc_{replicate}".format(replicate=replicate)
     fc_zscore_id = "zscore_" + fc_replicate_id
     empirical_bayes_id = "eb_std_{replicate}".format(replicate=replicate)
-    one_based_idx = replicate + 1
+    # one_based_idx = replicate + 1
```
Comment (mashehu) [inline bin/drugz.py:126]:
```suggestion
```

File: .pre-commit-config.yaml [hunk 10]
```diff
@@ -1,12 +1,10 @@
...
+  - repo: https://github.com/astral-sh/ruff-pre-commit
+    rev: v0.4.3
     hooks:
-      - id: black
-  - repo: https://github.com/pycqa/isort
-    rev: 5.12.0
-    hooks:
-      - id: isort
+      - id: ruff # linter
+        args: [--fix, --exit-non-zero-on-fix] # sort imports and fix
```
Comment (mashehu) [inline .pre-commit-config.yaml:6]:
```suggestion
        args: [--fix, --exit-non-zero-on-fix, --select E4,E7,E9,F,I] # default ruff rules + import sorting
```

**Review by mashehu — APPROVED** [review]

File: .pre-commit-config.yaml [hunk 11]
```diff
@@ -1,12 +1,10 @@
...
+  - repo: https://github.com/astral-sh/ruff-pre-commit
+    rev: v0.4.3
     hooks:
-      - id: black
-  - repo: https://github.com/pycqa/isort
-    rev: 5.12.0
-    hooks:
-      - id: isort
+      - id: ruff # linter
+        args: [--fix, --exit-non-zero-on-fix, --select E4,E7,E9,F,I] # default ruff rules + import sorting
```
Comment (LaurenceKuhl) [inline .pre-commit-config.yaml:6]:
```suggestion
        args: [--fix, --exit-non-zero-on-fix] # sort imports and fix
```

**Pull request merged**